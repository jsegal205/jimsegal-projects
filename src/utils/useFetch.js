import { useState, useEffect } from "react";
import axios from "axios";

// custom hook for performing GET request
const useFetch = (url, initialValue) => {
  const [data, setData] = useState(initialValue);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await axios.get(url);
        setData(response.data);
      } catch ({ response }) {
        setData({ status: response.status });
      } finally {
        setLoading(false);
      }
    };
    fetchData();
  }, [url]);
  return { loading, data };
};

export default useFetch;
