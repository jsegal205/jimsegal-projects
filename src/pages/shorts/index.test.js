import React from "react";
import { BrowserRouter as Router } from "react-router-dom";

import useFetch from "../../utils/useFetch";
import Shorts from "./index";
import { render } from "@testing-library/react";

jest.mock("../../utils/useFetch", () => jest.fn());

describe("Shorts Component", () => {
  describe("when `useFetch` is awaiting promise to resolve", () => {
    it("displays `Loading...`", () => {
      useFetch.mockReturnValue({
        loading: true,
        data: null,
      });

      const { container } = render(
        <Router>
          <Shorts />
        </Router>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe("when `useFetch` returns data", () => {
    it("displays data", () => {
      useFetch.mockReturnValue({
        loading: false,
        data: {
          probability: 50,
          criteria: [{ label: "criteria 1", value: "value 1" }],
        },
      });

      const { container } = render(
        <Router>
          <Shorts />
        </Router>
      );
      expect(container).toMatchSnapshot();
    });

    describe("when 500 status returned", () => {
      it("displays error messaging", () => {
        useFetch.mockReturnValue({
          loading: false,
          data: {
            error: "server likely on fire",
            status: 500,
          },
        });

        const { container } = render(
          <Router>
            <Shorts />
          </Router>
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
