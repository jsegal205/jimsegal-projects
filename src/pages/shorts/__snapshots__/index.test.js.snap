// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipes Component when \`useFetch\` is awaiting promise to resolve displays \`Loading...\` 1`] = `
<div>
  <section
    class="loader"
  >
    <div
      class="loader-wheel"
    />
    <div
      class="loader-text"
    >
      Loading
    </div>
  </section>
</div>
`;

exports[`Recipes Component when \`useFetch\` returns data displays data 1`] = `
<div>
  <section>
    <h2>
      Is Jim wearing shorts?
    </h2>
    <section>
      <article>
        There is a 
        <strong>
          50
          %
        </strong>
         chance he is.
      </article>
      <br />
      <article>
        Based on the following criteria:
        <ul>
          <li>
            criteria 1
            : 
            value 1
          </li>
        </ul>
      </article>
    </section>
  </section>
</div>
`;

exports[`Recipes Component when \`useFetch\` returns data when 500 status returned displays error messaging 1`] = `
<div>
  <section>
    <h3>
      Whoops! There was a problem loading
       
      "Is Jim Wearing Shorts" data.
    </h3>
    <p>
      Please reload browser to try again in a little bit.
    </p>
  </section>
</div>
`;
